test/cpp/end2end/time_change_test.cc

```diff
@@ -0,0 +1,364 @@
+    ts.tv_nsec -= GPR_NS_PER_SEC;
+    ++ts.tv_sec;
+  } else if (ts.tv_nsec < 0) {
+    --ts.tv_sec;
+    ts.tv_nsec = GPR_NS_PER_SEC + ts.tv_nsec;
+  }
+  return ts;
+}
+
+// offset the value returned by gpr_now(GPR_CLOCK_REALTIME) by msecs
+// milliseconds
+static void set_now_offset(int msecs) {
+  g_time_shift_sec = msecs / 1000;
+  g_time_shift_nsec = (msecs % 1000) * 1e6;
+}
+
+// restore the original implementation of gpr_now()
+static void reset_now_offset() {
+  g_time_shift_sec = 0;
+  g_time_shift_nsec = 0;
+}
+
+namespace grpc {
+namespace testing {
+
+namespace {
+
+// gpr_now() is called with invalid clock_type
+TEST(TimespecTest, GprNowInvalidClockType) {
+  // initialize to some junk value
+  gpr_clock_type invalid_clock_type = (gpr_clock_type)32641;
+  EXPECT_DEATH(gpr_now(invalid_clock_type), ".*");
+}
+
+// Add timespan with negative nanoseconds
+TEST(TimespecTest, GprTimeAddNegativeNs) {
+  gpr_timespec now = gpr_now(GPR_CLOCK_MONOTONIC);
+  gpr_timespec bad_ts = {1, -1000, GPR_TIMESPAN};
+  EXPECT_DEATH(gpr_time_add(now, bad_ts), ".*");
+}
+
+// Subtract timespan with negative nanoseconds
+TEST(TimespecTest, GprTimeSubNegativeNs) {
+  // Nanoseconds must always be positive. Negative timestamps are represented by
+  // (negative seconds, positive nanoseconds)
+  gpr_timespec now = gpr_now(GPR_CLOCK_MONOTONIC);
+  gpr_timespec bad_ts = {1, -1000, GPR_TIMESPAN};
+  EXPECT_DEATH(gpr_time_sub(now, bad_ts), ".*");
+}
+
+// Add negative milliseconds to gpr_timespec
+TEST(TimespecTest, GrpcNegativeMillisToTimespec) {
+  // -1500 milliseconds converts to timespec (-2 secs, 5 * 10^8 nsec)
+  gpr_timespec ts = grpc_millis_to_timespec(-1500, GPR_CLOCK_MONOTONIC);
+  GPR_ASSERT(ts.tv_sec = -2);
+  GPR_ASSERT(ts.tv_nsec = 5e8);
+  GPR_ASSERT(ts.clock_type == GPR_CLOCK_MONOTONIC);
+}
+
+class TimeChangeTest : public ::testing::Test {
+ protected:
+  TimeChangeTest() {}
+
+  void SetUp() {
+    auto port = grpc_pick_unused_port_or_die();
+    std::ostringstream addr_stream;
+    addr_stream << "localhost:" << port;
+    auto addr = addr_stream.str();
+    server_.reset(new SubProcess({
+        g_root + "/client_crash_test_server",
+        "--address=" + addr,
+    }));
+    GPR_ASSERT(server_);
+    channel_ = CreateChannel(addr, InsecureChannelCredentials());
+    GPR_ASSERT(channel_);
+    stub_ = grpc::testing::EchoTestService::NewStub(channel_);
+  }
+
+  void TearDown() {
+    server_.reset();
+    reset_now_offset();
+  }
+
+  std::unique_ptr<grpc::testing::EchoTestService::Stub> CreateStub() {
+    return grpc::testing::EchoTestService::NewStub(channel_);
+  }
+
+  std::shared_ptr<Channel> GetChannel() { return channel_; }
+
+  std::unique_ptr<SubProcess> server_;
+  std::shared_ptr<Channel> channel_;
+  std::unique_ptr<grpc::testing::EchoTestService::Stub> stub_;
+};
+
+// Wall-clock time jumps forward on client before bidi stream is created
+TEST_F(TimeChangeTest, TimeJumpForwardBeforeStreamCreated) {
```

Comment:
Is it possible to produce multiple shifts in a test? Just more extreme cases that worth being tested and also covers the one-shift case.

Reply:
Added a new test case `TimeJumpForwardAndBackDuringCall` with multiple shifts during a call.